core/commonKotlin/test/InstantIsoStringsTest.kt

```diff
@@ -0,0 +1,420 @@
+            "1970-02-03T04:05:06.123456789+01:2:60",
+            "1970-02-03T04:05:06.123456789+01:12:6",
+        )) {
+            assertInvalidFormat(nonIsoString) { parseInstant(nonIsoString) }
+        }
+        // this string represents an Instant that is currently larger than Instant.MAX any of the implementations:
+        assertInvalidFormat { parseInstant  ("+1000000001-12-31T23:59:59.000000000Z") }
+    }
+
+    @Test
+    fun parseStringsWithOffsets() {
+        val strings = arrayOf(
+            Pair("2020-01-01T00:01:01.02+18:00", "2019-12-31T06:01:01.020Z"),
+            Pair("2020-01-01T00:01:01.123456789-17:59:59", "2020-01-01T18:01:00.123456789Z"),
+            Pair("2020-01-01T00:01:01.010203040+17:59:59", "2019-12-31T06:01:02.010203040Z"),
+            Pair("2020-01-01T00:01:01.010203040+17:59", "2019-12-31T06:02:01.010203040Z"),
+            Pair("2020-01-01T00:01:01+00", "2020-01-01T00:01:01Z"),
+        )
+        strings.forEach { (str, strInZ) ->
+            val instant = parseInstant(str)
+            assertEquals(parseInstant(strInZ), instant, str)
+            assertEquals(strInZ, displayInstant(instant), str)
+        }
+        assertInvalidFormat { parseInstant("2020-01-01T00:01:01+18:01") }
+        assertInvalidFormat { parseInstant("2020-01-01T00:01:01+1801") }
+        assertInvalidFormat { parseInstant("2020-01-01T00:01:01+0") }
+        assertInvalidFormat { parseInstant("2020-01-01T00:01:01+") }
+        assertInvalidFormat { parseInstant("2020-01-01T00:01:01") }
+        assertInvalidFormat { parseInstant("2020-01-01T00:01:01+000000") }
+
+        val instants = listOf(
+            Instant.DISTANT_FUTURE,
+            Instant.DISTANT_PAST,
+            Instant.fromEpochSeconds(0, 0),
+            Instant.parse("2020-01-02T03:04:05.6789Z"),
+            Instant.MAX,
+            Instant.MIN,
+        )
+
+        val offsets = listOf(
+            0 to "Z",
+            3 * 3600 + 12 * 60 + 14 to "+03:12:14",
+            - 3 * 3600 - 12 * 60 - 14 to "-03:12:14",
+            2 * 3600 + 35 * 60 to "+02:35",
+            - 2 * 3600 - 35 * 60 to "-02:35",
+            4 * 3600 to "+04",
+            - 4 * 3600 to "-04",
+        )
+
+        for (instant in instants) {
+            for ((offsetSeconds, offsetString) in offsets) {
+                if (instant == Instant.MAX && offsetSeconds < 0 ||
+                    instant == Instant.MIN && offsetSeconds > 0
+                ) continue
+                val newInstant = Instant.parse("${instant.toString().dropLast(1)}$offsetString")
+                assertEquals(newInstant, instant.minus(offsetSeconds.seconds))
+            }
+        }
+    }
+
+    private fun parseInstant(isoString: String): Instant {
+        // return Instant.parse(isoString)
+        return parseIso(isoString)
+    }
+
+    private fun displayInstant(instant: Instant): String {
+        // return instant.toString()
+        return formatIso(instant)
+    }
```

Comment:
@qurbonzoda, here's the logic behind this PR: the newly-added `formatIso` can replace `toString`, and the newly-added `parseIso` can replace `Instant.parse` in the case when no specific format is passed (which is the only case supported by `kotlin.time.Instant).